File: unionfind.py
# Union-Find
class UnionFind():
	def __init__(self, n: int) -> None:
		self.n = n
		self.par = list(range(self.n))
		self.rank = [1] * self.n
		self.count = self.n

	def find(self, x: int) -> int:
		if self.par[x] == x:
			return x
		else:
			self.par[x] = self.find(self.par[x])
			return self.par[x]

	def unite(self, x: int, y: int) -> None:
		p = self.find(x)
		q = self.find(y)
		if p == q:
			return None
		if p > q:
			p, q = q, p
		self.rank[p] += self.rank[q]
		self.par[q] = p
		self.count -= 1

	def same(self, x: int, y: int) -> bool:
		return self.find(x) == self.find(y)

	def size(self, x: int) -> int:
		return self.rank[x]

	def count(self) -> int:
		return self.count

File: Cpp/data_structure/union_find.cpp
#include <iostream>
#include <cassert>
#include <vector>

struct union_find {
	public:
		union_find(int size) {
			n = size;
			parent.resize(n);
			rank.resize(n);
			for (int i = 0; i < n; i++) {
				parent[i] = i;
				rank[i] = 1;
			}
			cnt = n;
		}

		int find(int x) {
			assert(0 <= x && x < n);
			if (parent[x] == x) return x;
			return parent[x] = find(parent[x]);
		}

		int unite(int x, int y) {
			assert(0 <= x && x < n);
			assert(0 <= y && y < n);
			int p = find(x), q = find(y);
			if (p == q) return p;
			if (p > q) std::swap(p, q);
			rank[p] += rank[q];
			parent[q] = p;
			cnt -= 1;
			return p;
		}
	
	bool same(int x, int y) {
		assert(0 <= x && x < n);
		assert(0 <= y && y < n);
		return find(x) == find(y);
	}

	int size(int x) {
		assert(0 <= x && x < n);
		return rank[x];
	}

	int count() {
		return cnt;
	}

	private:
		int n, cnt;
		std::vector<int> parent, rank;
};

int main() {
	int n, q, t, u, v;
	std::cin >> n >> q;
	union_find uf(n);
	for (int i = 0; i < q ; i++) {
		std::cin >> t >> u >> v;
		if (t == 0) uf.unite(u, v);
		else {
			if (uf.same(u, v)) std::cout << 1 << std::endl;
			else std::cout << 0 << std::endl;
		}
	}
}